### PR TITLE
[on ice] Use AIM for trading

### PR DESCRIPTION
**Pull request by andrei8l**
#### Summary

`SUMMARY: Features "Implement a trade mode for AIM and allow using it for trading with NPCs"`

#### Purpose of change
The NPC trading UI is less developed and functional than every other interface in CDDA. Instead of duplicating work to develop it further, we can re-use the featureful Advanced Inventory Manager.
Related to: https://github.com/CleverRaven/Cataclysm-DDA/issues/42363

#### Describe the solution
Implement a trade mode in the Advanced Inventory Manager. This is disabled by default and can be enabled from global settings -> debug.
While in trade mode:
1. header shows participants and balance instead of message log
2. left pane shows the NPC's items, right pane shows PC's items
2. items are referenced in a tradelist  (limbo) instead of being transferred directly.
3. if the NPC is a shopkeeper, the [A] square in their pane shows all owned items in `PICKUP_RANGE`
4. a new square [T] shows the tradelist for that party (outgoing items)
5. trying to exit the AIM will do the usual balance checks and transfer the items if the trade is accepted by both parties. There is also a new action to cancel the trade and exit quickly

Should have feature parity with `npc_trading`, unless I missed something.

known defects:
1. items are transferred instantly (no move cost) - same as for `npc_trading`
2. allows for item teleportation within NPC interaction range (+PICKUP_RANGE for shopkeepers) - same as for `npc_trading`
3. `can_stash` check is only done for each stack rather than for the whole tradelist - this is pretty broken for `npc_trading` and even PC pickup as well. A fugly solution would be to put a copy of the tradelist in a fake container and run `can_stash` on that
4. ~~squares are centered on the PC even in the NPC pane (except for [A] for a shopkeeper) - I couldn't figure out a clean solution for this~~ FIXED in https://github.com/CleverRaven/Cataclysm-DDA/pull/42748/commits/2dfc7106e675a17ea4413e2a7e891df39ba55d7b

1-3 are largely out of scope for this PR. ~~I'd like to find a workaround or solution for 4 but I haven't had any good ideas yet.~~

minor improvements over `npc_trading`:
1. the actual items from a stack are transferred instead of copies of the first element
2. actually trains barter skill

#### Describe alternatives you've considered
N/A

#### Testing
Trade various items with an NPC shopkeeper (free trader and intercom from Hub 01), an allied NPC, and a non-allied not-shopkeep NPC and confirm that the appropriate amounts get transferred.
Move items to/from tradelist while adjacent to a few thousand items to check performance impact
Verify that item lists and counts are updated properly when stacks/charges are moved to/from the tradelist
Verify that AIM still functions correctly in regular mode.

Mostly tested, but there are probably still edge cases.
I've tested every feature as I wrote it, but this thing grew into a pretty ugly mess and it's very possible I broke something (or missed it entirely)
~~I didn't test at all after merging master. TODO~~

#### Additional context
I copied some code (with modifictions) from `npc_trading` because that was easier than refactoring the old code.
I had to make an invasive change in `item::display_name()` to make the interface work with `count_by_charges` items
I moved `get_AIM_inventory` out of `avatar` and ~~into an anonymous namespace~~ made it static because it really had no purpose in `avatar`

### CODE REVIEW

**Comment by Fris0uman** [comment]
Can you post some screen shots of the thing?

**Comment by andrei8l** [comment]
Only the header changed
Untitled

**Comment by andrei8l** [comment]
`mx_minefield real spawn` my eternal bane

**Comment by Night-Pryanik** [comment]
We probably need to add a `!nofail` parameter for this test.

File: src/npctrade.cpp [hunk 1]
```diff
@@ -674,6 +676,10 @@ bool npc_trading::trade( npc &np, int cost, const std::string &deal )
     //               np.volume_carried() - np.volume_capacity() );
     np.drop_invalid_inventory();
 
+    if( get_option<bool>( "AIM_TRADE" ) ) {
```
Comment (mlangsdorf) [inline src/npctrade.cpp:679]: Can we pass the `deal` string to create_advanced_inv and include it in the header?

Comment (andrei8l) [reply src/npctrade.cpp:679]: Fix in latest commit. Thanks for pointing this out

**Comment by andrei8l** [comment]
> Aside from the slightly snide PR description, this seems fine on first glance and if it works at all, is definitely an improvement over the current trading interface that got broken by nested containers (and wasn't great to begin with).

Fixed; sorry. I was going for funny rather than snide.

> 
> I don't know enough about AIM to review.
> 
> Also, since the trading window is currently broken, this code should not be gated behind a debug option, but should entirely replace the code in npctrade.cpp.

I'd like to keep `npc_trading` around for now since it mostly works and people are used to it. If this gets merged, the maintainer can flip that bool if this UI is preferred.

**Comment by andrei8l** [comment]
I'd like to get this merged in 0.F if possible. It addresses one of the blockers and can be disabled if it turns out to be problematic. It would also help with testing so we can eventually make this the default trading UI.

**Comment by ZhilkinSerg** [comment]
> 
> 
> I'd like to get this merged in 0.F if possible. It addresses one of the blockers and can be disabled if it turns out to be problematic. It would also help with testing so we can eventually make this the default trading UI.

Don't mind release blocker tags yet - they are not really set in stone.

**Comment by andrei8l** [comment]
> While AIM is very feature-full, it's also very complicated and the code is a mess internally, shimming more functionality into it is very problematic from a maintainability point of view. For that reason alone I do not want anything resembling this PR to go into 0.F, this is exactly the kind of risky set of changes we do not want right now. What I'd want to see before expanding the scope of AIM still further is for it to be refactored to pull some of the crazy functionality out and make it easier to maintain. If we keep adding things on top of it, it's just going to be harder and harder to do that, and it's already prohibitively difficult.

Ok.  Please elaborate on the crazy functionality and I can look into it.

> 
> As for the overall approach even excepting that, this:
> 
> ```
> 
>     the actual items from a stack are transferred instead of copies of the first element
>     actually trains barter skill
> ```
> 
> Doesn't seem like remotely sufficient justification for taking this route, even if AIM wasn't so complicated. You might have simply not enumerated the improvements, but as it is this is a very weak case for re-using AIM, it and trading have almost nothing to do with each other.

The AIM is a well developed UI for listing items and is very well suited for trading. I did not think I needed to enumerate its advantages over an unsorted, unstacked, non-interactive list of items (a.k.a. `npc_trading`). The two things mentioned above were just minor improvements over bugs I noticed while developing this. They are not the selling points.

> As for the option, that's a hard no, you've either fixed it or you haven't, we definitely aren't going to drag around two implementations of the trading interface just in case one of them is missing some features or has some bugs.

Ok... I tried the conservative approach but this can't be accepted. Should I have slashed 500 lines of code out of `npc_trading` in the same PR? This code does work and it does not change the base functionality of the AIM.

I am not certain on what changes you're requesting. It kinda seems like you're asking me to wait around until someone else refactors the AIM.

**Comment by I-am-Erk** [comment]
So I'm not a coder and only understand that any time I suggest a fix to AIM I get a lot of thousand yard stares.

I love what you're going for here and I'd really like to see someone working on AIM and on trading. I think it's one of our bett Inventory systems and ideally, should be integrated with plain inventory management. Using it for trade is kind of a stroke of genius in many ways.

That said, the problem here isn't what you're doing, it's what the rest of us do once you're gone. The AIM code is a mess and nobody knows it. When your work bugs or needs changes, we're hooped.

If you want this to be integrated - which I wholeheartedly support as a concept - then it *requires* the backing framework of a refactor to AIM to make it so that other people can work on this going forward. This is something we all want deeply, so while I realize that's a kick in the gut when you've made something excellent, I hope you'd seriously consider doing it and then implementing this.

**Comment by andrei8l** [comment]
> That said, the problem here isn't what you're doing, it's what the rest of us do once you're gone.

Are you expecting or hoping me to leave so early? I've only just joined.

> I hope you'd seriously consider doing it and then implementing this.

Sure; just let me know what it needs done, in more specific terms. I only noticed a couple of things wrong with it and they do not require massive refactors (namely: the monstrous move_all action and duplicated checks with activity actor)

**Comment by kevingranade** [comment]
> The AIM is a well developed UI for listing items and is very well suited for trading.

We have a number of menus that are good at listing items, AIM is massively over engineered for choosing a set of items to move from point A to point B unless you need to arbitrarily change what point A and point B are on the fly, which you don't need to do when trading.
I would also disagree with "well developed", it's a sprawling and complicated body of code.  If it were a lot more modular such that the new code weren't embedded in the already overcomplicated code, sure it'd be fine, but it's not, every additional feature drives the complexity of the code up.

> Should I have slashed 500 lines of code out of npc_trading in the same PR?

Yes, ish.  If you are replacing functionality, the old functionality needs to go, the old 500-odd lines of code represents a maintenance burden that we cannot afford.  However other problems with this approach means that this wouldn't suddenly make the PR mergeable.

> It kinda seems like you're asking me to wait around until someone else refactors the AIM.

Until either you or someone else does it, yes.

> Are you expecting or hoping me to leave so early? I've only just joined.

I expect *any given developer* to leave at the drop of a hat, leaving any partially complete features in whatever state they were in at the end of their latest PR, because it happens.  As such, I don't accept PRs that leave the code in a bad state if I can at all help it.  I cut a little more slack to developers with a track record of following through on changes and that I'm currently in close contact with, but this particular set of changes wouldn't be a good candidate for that in any event.

There are two directions you could take AIM, and I haven't dug into it to figure out which one is better.
One is to *extract* the business logic of picking up and dropping items, overhauling AIM to accept them as arguments when it's invoked, this is how the code in ui.h tends to work.
The other option is to fragment AIM into a bunch of helpers than can be invoked by its callers in order to build a window instance.

**Comment by andrei8l** [comment]
> We have a number of menus that are good at listing items, AIM is massively over engineered for choosing a set of items to move from point A to point B unless you need to arbitrarily change what point A and point B are on the fly, which you don't need to do when trading.

It does a fair bit more than that, in particular listing by category and arbitrary sorting. Having separate squares for loot is useful in trading too (consider "loot I want to sell" and "loot I might sell if the trader has good offers"), or equipping followers with distinct sets of gear without having to run around out of range of certain items.

> I would also disagree with "well developed", it's a sprawling and complicated body of code.

I rather meant in terms of features for listing items. There is nothing else that comes even close.

> I expect _any given developer_ to leave at the drop of a hat, leaving any partially complete features in whatever state they were in at the end of their latest PR, because it happens.

Ok, fair.

> 
> There are two directions you could take AIM, and I haven't dug into it to figure out which one is better.
> One is to _extract_ the business logic of picking up and dropping items, overhauling AIM to accept them as arguments when it's invoked, this is how the code in ui.h tends to work.

The business logic of picking up and dropping items is already split to activity actors.

> The other option is to fragment AIM into a bunch of helpers than can be invoked by its callers in order to build a window instance.

So turn it into a UI library?

**Comment by I-am-Erk** [comment]
We're not expecting you to leave soon, but it's sort of irrelevant. In a project like this we try to treat every PR as if there's a high chance the person writing it will just disappear soon, because that's what happens.

I wrote this a bit ago and it seems other responses have passed me by before I can finish the thought, but based on my end of understanding I'd like to see the concepts of AIM split up into a modular UI concept that could be invoked from the main Inventory to get group moving and things, or alternatively called from the NPC end when, say, a trade interaction is invoked. These should both draw on the core UI framework that is AIM and have a consistent set of features, but not be functionally the same element.

**Comment by andrei8l** [comment]
Alright. You're only asking for the moon so let me see what I can do. It'll likely take some time though, so please be patient.

**Comment by I-am-Erk** [comment]
Your ongoing efforts to find and secure the moon are deeply appreciated, and will make my own Moon Laser Project exponentially easier.

**Comment by andrei8l** [comment]
I haven't really had the time to work on this and I've noticed that someone already tried to rewrite the AIM recently but that PR was also rejected, so my interest in this isn't very high atm. I still plan to return to this code sometime.

**Comment by andrei8l** [comment]
In case you're still waiting for this: I've finally started working on rewriting the AIM with generalized UI elements. I hope to have something ready by next week (trade mode would come in a subsequent PR). I'm not promising anything though.

**Comment by I-am-Erk** [comment]
Awesome, glad to hear it. Hope it gets somewhere. I'd love AIM to be something we integrate with more things and that requires us to have code people can work with
